### PR TITLE
Move CPROVER_{r,w,rw}_ok processing to back-end

### LINKED ISSUE #7377: Silent transformation of __CPROVER_{r,w,rw}_ok causing surprising behaviour
Posting an analysis by @nwetzler, @jimgrundy and @remi-delmas-3000  The [attached a gzipped tarball](https://github.com/diffblue/cbmc/files/10076605/goto-instrument-pointer-checks.tar.gz) contains a single C file and a script to demonstrate the problem.
 
The following function takes a single pointer and asserts that it is read/write okay:
```
void foo (char* x) {
    __CPROVER_assert(__CPROVER_rw_ok(x, 1), "");
    return;
}
```
 
We can compile this to goto (to create a.out) and then run goto-instrument on the resulting binary to create a new binary (which is b.out).  Off the bat, you’ll notice that the size of these two binaries is different.  However, if you show-goto-functions or program-only (for the SSA) and compare the two, they will be identical.
 
But if you now enable pointer-check and pointer-primitive-check, you’ll see that they produce different show-goto-functions and program-only output.  Furthermore, we think the assertions produced by the checks are even semantically different between the two, which seems like a bug in CBMC.

```
83c83
<         ASSERT pointer_object(NULL) = pointer_object(foo::x) ∨ ¬(is_invalid_pointer(foo::x)) // pointer invalid in RW_OK(x, (unsigned long int)1)
---
>         ASSERT (¬(pointer_object(foo::x) = pointer_object(NULL)) ∧ ¬(is_invalid_pointer(foo::x)) ∧ ¬(pointer_object(foo::x) = pointer_object(__CPROVER_deallocated)) ∧ ¬(pointer_object(foo::x) = pointer_object(__CPROVER_dead_object)) ∧ pointer_offset(foo::x) ≥ 0) ⇒ (pointer_object(NULL) = pointer_object(foo::x) ∨ ¬(is_invalid_pointer(foo::x))) // pointer invalid in OBJECT_SIZE(x)
```

Several issues appear to arise:
1. There isn't actually a way to just dump the contents of a GOTO binary for `goto-instrument` will always call into instrumentation steps.
2. Just like you cannot just dump a GOTO binary, running `goto-instrument a.out b.out` will perform transformations despite no such transformation being requested.
3. A particular example of such a transformation (maybe the only one, maybe there are others) is the expansion of `__CPROVER_{r,w,rw}_ok` that `goto_check_ct` does (irrespective of whether a property instrumentation was requested).
4. `goto_check_ct` is not idempotent: doing one round of `goto_check_ct` followed by another with `--pointer-primitive-check` enabled is not the same as doing just one round of `goto_check_ct` with `--pointer-primitive-check` enabled.

### PR DESCRIPTION
**Pull request by tautschnig**
Please review commit-by-commit.

Fixes https://github.com/diffblue/cbmc/issues/7377.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/util/pointer_expr.cpp
Comment (remi-delmas-3000) [inline src/util/pointer_expr.cpp]: Would these expressions be lowered in an ulterior lowering step ? (just thinking about what happen when passing these through `goto-instrument` multiple times).

EDIT: just saw that `same_object` is a function returning a more basic expression and not an `exprt`  subtype, you can disregard my comment.

File: src/util/pointer_expr.cpp
Comment (remi-delmas-3000) [inline src/util/pointer_expr.cpp]: This condition was not present in `good_pointer_def` which `r_or_rw_ok` is now supposed to replace (cf https://github.com/diffblue/cbmc/pull/7386/files)

File: src/util/pointer_expr.h
Comment (remi-delmas-3000) [inline src/util/pointer_expr.h:501]: is this method inherited from some superclass  or trait, or is it added on a per-class basis ?

File: src/solvers/flattening/bv_pointers.cpp
Comment (remi-delmas-3000) [inline src/solvers/flattening/bv_pointers.cpp]: It seems we don't have any test that triggers this lowering step. Do we really get the guarantee that rw_ok expressions are going to be converted only when going to the solver ? (assuming this is what this PR tries to achieve ?)

File: src/solvers/smt2/smt2_conv.cpp
Comment (remi-delmas-3000) [inline src/solvers/smt2/smt2_conv.cpp]: no coverage here,  could this be a sign that these expressions still get lowered before reaching that step ?

**Comment by jimgrundy** [comment]
Looks like we need to add __CPROVER_is_in_range to the mix

**Comment by tautschnig** [comment]
> Looks like we need to add __CPROVER_is_in_range to the mix

This is done (including a few fixes specific to just that).

File: src/solvers/smt2/smt2_conv.cpp
Comment (tautschnig) [reply src/solvers/smt2/smt2_conv.cpp]: This should now be ok - at the time of you viewing this I had not included the removal of early expansion.

File: src/solvers/flattening/bv_pointers.cpp
Comment (tautschnig) [reply src/solvers/flattening/bv_pointers.cpp]: Should now be done, the early-expansion removal wasn't there back then.

File: src/util/pointer_expr.cpp
Comment (tautschnig) [reply src/util/pointer_expr.cpp]: I believe this was a bug in `good_pointer_def`.

File: src/ansi-c/c_typecheck_expr.cpp
Comment (kroening) [inline src/ansi-c/c_typecheck_expr.cpp]: There's something wrong here. The connection to `deallocated` is very much a mechanism for implementing the predicate, and that shouldn't happen in the front-end.

**Review by kroening — CHANGES_REQUESTED** [review]
There's a fundamental architectural problem with this PR.
Just imagine you had to write an SMT-LIB standard proposal for a theory like this.

**Comment by tautschnig** [comment]
> There's a fundamental architectural problem with this PR. Just imagine you had to write an SMT-LIB standard proposal for a theory like this.

I agree that there is an architectural problem, but I'm not convinced it's in this PR: this PR just makes it very obvious: as they are (before this PR), `{r,w,rw}_ok_exprt` and `pointer_in_range_exprt` are not actually expressions: two occurrences with the same set of (previous) arguments may still result in different evaluations. Such "expressions" we would otherwise prefix with `side_effect_`.

File: src/ansi-c/c_typecheck_expr.cpp
Comment (tautschnig) [reply src/ansi-c/c_typecheck_expr.cpp]: Well, I really just moved (while also making it more obvious) something from `goto_check_ct` to type checking. Both of which are in the front-end...

**Comment by jimgrundy** [comment]
I'm in agreement with Michael. The prior API aligns to the behavior one wishes these primitives had, but not the behavior they actually have. This dependence on the DEAD and DEALLOCATED values is not desirable, but also very real, and the way it is obscured by the present API has proved confusing. I think it best that the API make the behavior plain.

That's not to say that we wouldn't one day aspire to improving the behavior, perhaps by using SMT more in the encoding, and at that time setting the API back to what it was before. I'd very much like to see that.

File: src/ansi-c/c_typecheck_expr.cpp
Comment (peterschrammel) [reply src/ansi-c/c_typecheck_expr.cpp]: Would it help if we lowered the side-effecty function to a predicate in the built-in library?
The C implementation of the function can refer to the global state to call the built-in predicate.

**Comment by thomasspriggs** [comment]
I'd prefer it if we can avoid removing the `new-smt-backend` tag from existing tests due to broken functionality. We should be able to add the required functionality to the incremental decision procedure via the following steps -

1. Take the code introduced into `smt2_conv.cpp` and separate it into its own function, so that it can be used by both decision procedures.
2. Add calls to the new function in the two places in `smt2_incremental_decision_procedure.cpp` where `lower_byte_operators` is currently called.

**Comment by tautschnig** [comment]
> I'd prefer it if we can avoid removing the `new-smt-backend` tag from existing tests due to broken functionality. We should be able to add the required functionality to the incremental decision procedure via the following steps -
> 
>     1. Take the code introduced into `smt2_conv.cpp` and separate it into its own function, so that it can be used by both decision procedures.
> 
>     2. Add calls to the new function in the two places in `smt2_incremental_decision_procedure.cpp` where `lower_byte_operators` is currently called.

Thank you for the guidance, I'll implement this.

**Comment by tautschnig** [comment]
> There's a fundamental architectural problem with this PR. Just imagine you had to write an SMT-LIB standard proposal for a theory like this.

@kroening Would you mind taking another look at the implementation, specifically in the commit [r/w/rw_ok and pointer_in_range depend on deallocated and dead_object](https://github.com/diffblue/cbmc/pull/7395/commits/1ba2ee65c85958d414109e988f82e3409db795fc)?

File: src/goto-programs/builtin_functions.cpp
Comment (kroening) [inline src/goto-programs/builtin_functions.cpp]: Do you still need this change?

**Comment by kroening** [comment]
I have on remaining problem with this PR.

There are now two variants of the memory predicates: prior to filling in the particular realisation using `dead` and `deallocated`, and post. They are very different, and in particular, there may well be many different realisations.

May I thus advise to make this separate expression classes, with different arities, and different identifiers -- much like the trick used in the CHC encoder. Right now this is very difficult to understand by looking at the code, or the classes.

File: src/goto-programs/builtin_functions.cpp
Comment (tautschnig) [reply src/goto-programs/builtin_functions.cpp]: It won’t be subject to cleaning at this point, but I could of course explicitly invoke cleaning.

**Comment by tautschnig** [comment]
> May I thus advise to make this separate expression classes, with different arities, and different identifiers -- much like the trick used in the CHC encoder. Right now this is very difficult to understand by looking at the code, or the classes.

Any preference on renaming the front end vs the to-be-back-end ones? We could have ones with a “frontend” in their name like we do for some codet classes.

**Comment by kroening** [comment]
> Any preference on renaming the front end vs the to-be-back-end ones? We could have ones with a “frontend” in their name like we do for some codet classes.

We have three categories of expressions:
1. Ones that only exist in some of the language front-ends, but never make their way into the GOTO binary. We have used the`frontend_` prefix for those.
2. Ones that exist in the GOTO binary, and may in fact have an implicit "program state" argument.
3. Proper logic ones. No program state.

The one without the extra arguments fall into category 2. Once you add the arguments you have an expression in category 3. I'd use `r/w_ok` (no prefix) for the ones prior to baking in the state. We have sometimes used the prefix `mathematical_` to identify expressions and types of the third kind, and to distinguish these from their program counterparts.

The variant with the extra arguments is a specific realisation of the predicates that uses prophecy variables. How about prefixing them with `prophecy_`.

**Comment by martin-cs** [comment]
@tautschnig I still have a "review requested" for this.  Given the discussion and the conflicting files, I will rely on you saying when you want it reviewed.

**Comment by tautschnig** [comment]
> > I'd prefer it if we can avoid removing the `new-smt-backend` tag from existing tests due to broken functionality. We should be able to add the required functionality to the incremental decision procedure via the following steps -
> > ```
> > 1. Take the code introduced into `smt2_conv.cpp` and separate it into its own function, so that it can be used by both decision procedures.
> > 
> > 2. Add calls to the new function in the two places in `smt2_incremental_decision_procedure.cpp` where `lower_byte_operators` is currently called.
> > ```
> 
> Thank you for the guidance, I'll implement this.

This is now done.

**Comment by tautschnig** [comment]
> The variant with the extra arguments is a specific realisation of the predicates that uses prophecy variables. How about prefixing them with `prophecy_`.

I believe this is now implemented as suggested. Could I please get a re-review?

**Comment by thomasspriggs** [comment]
> > > I'd prefer it if we can avoid removing the `new-smt-backend` tag from existing tests due to broken functionality. We should be able to add the required functionality to the incremental decision procedure via the following steps -
> > > ```
> > > 1. Take the code introduced into `smt2_conv.cpp` and separate it into its own function, so that it can be used by both decision procedures.
> > > 
> > > 2. Add calls to the new function in the two places in `smt2_incremental_decision_procedure.cpp` where `lower_byte_operators` is currently called.
> > > ```
> > 
> > 
> > Thank you for the guidance, I'll implement this.
> 
> This is now done.

Your branch is still removing the `new-smt-backend` from three `.desc` files. Can you remove the removal from your PR please? I tried this myself locally and can confirm that these tests now pass with the lowering you added. My branch for this (showing the three tests) is here - https://github.com/thomasspriggs/cbmc/commit/9780929a26f773f136749c1184168d7dd95eb039

The changes you made to the `smt2_incremental_decision_procedure` look fine to me. Just FYI myself and @esteffin are each working on changes which are likely to touch the same lines of code where the decision procedure does the lowering. So there may be a need for some minor rebasing to resolve merge conflicts between the 3 of us in the near future.

File: src/ansi-c/expr2c_class.h
Comment (thomasspriggs) [inline src/ansi-c/expr2c_class.h]: :pick: This change doesn't seem to match the commit message, as the commit message states it is adding a method, but the change is forward declaring a class.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: Nitpick: With the lowing added to the decision procedure, it is not so much that this is an unimplemented feature but that that should be `UNREACHABLE_BECAUSE("prophecy_r_or_w_ok expression should have been lowered by the decision procedure before conversion to smt terms.")`

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp
Comment (thomasspriggs) [inline src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: See previous `UNIMPLEMENTED_FEATURE` nitpick.

**Comment by tautschnig** [comment]
> Your branch is still removing the `new-smt-backend` from three `.desc` files. Can you remove the removal from your PR please? I tried this myself locally and can confirm that these tests now pass with the lowering you added. My branch for this (showing the three tests) is here - [thomasspriggs@9780929](https://github.com/thomasspriggs/cbmc/commit/9780929a26f773f136749c1184168d7dd95eb039)

These were failing in two of the regression tests only. I'll happily try this again, but there may be a need for more debugging.

**Comment by thomasspriggs** [comment]
> > Your branch is still removing the `new-smt-backend` from three `.desc` files. Can you remove the removal from your PR please? I tried this myself locally and can confirm that these tests now pass with the lowering you added. My branch for this (showing the three tests) is here - [thomasspriggs@9780929](https://github.com/thomasspriggs/cbmc/commit/9780929a26f773f136749c1184168d7dd95eb039)
> 
> These were failing in two of the regression tests only. I'll happily try this again, but there may be a need for more debugging.

:thinking: All tests passed on my local machine. I have raised this separate PR for CI purposes only, so I can see the results across all CI jobs - https://github.com/diffblue/cbmc/pull/7730

**Comment by tautschnig** [comment]
> 🤔 All tests passed on my local machine. I have raised this separate PR for CI purposes only, so I can see the results across all CI jobs - #7730

Looks like this is producing the failures I had seen: check-macos-12-cmake-clang and codecov-coverage-report are failing.

**Comment by thomasspriggs** [comment]
> > thinking All tests passed on my local machine. I have raised this separate PR for CI purposes only, so I can see the results across all CI jobs - #7730
> 
> Looks like this is producing the failures I had seen: check-macos-12-cmake-clang and codecov-coverage-report are failing.

I see. I hadn't realised that this was only failing on some platforms.

**Comment by thomasspriggs** [comment]
> > > thinking All tests passed on my local machine. I have raised this separate PR for CI purposes only, so I can see the results across all CI jobs - #7730
> > 
> > 
> > Looks like this is producing the failures I had seen: check-macos-12-cmake-clang and codecov-coverage-report are failing.
> 
> I see. I hadn't realised that this was only failing on some platforms.

Ok, it looks like removing the `simplify_expr` step fixes the 2 failing tests of the mac job - https://github.com/diffblue/cbmc/pull/7730/commits/79f95f8e43e7341dc1a11d04f889f03a22cc3c18 I have a hunch that the simplification is simplifying away expressions which need to be defined within the decision procedure for the `get` function to work as it should do, but I don't have a mac on hand to confirm right now.

That means the issue with `dynamic_sizeof1` and the coverage job is still a mystery for the moment.

**Comment by tautschnig** [comment]
> > > thinking All tests passed on my local machine. I have raised this separate PR for CI purposes only, so I can see the results across all CI jobs - #7730
> > 
> > 
> > Looks like this is producing the failures I had seen: check-macos-12-cmake-clang and codecov-coverage-report are failing.
> 
> I see. I hadn't realised that this was only failing on some platforms.

I believe I have root-caused this to https://github.com/diffblue/cbmc/pull/7735 (~~which is now also included in this PR~~ which is now merged).

**Comment by kroening** [comment]
This needs another nudge. At the moment, the lowering of X_ok to prophecy_X_ok is done during goto-conversion, which precludes any other approach. This should be done by a separate pass, appropriate for the analyser.

**Comment by kroening** [comment]
The contract instrumenter in goto-instrument should then generate X_ok, not prophecy_X_ok.

**Comment by tautschnig** [comment]
> This needs another nudge. At the moment, the lowering of X_ok to prophecy_X_ok is done during goto-conversion, which precludes any other approach. This should be done by a separate pass, appropriate for the analyser.

I have now moved the lowering to a separate pass called `rewrite_rw_ok`.

> The contract instrumenter in goto-instrument should then generate X_ok, not prophecy_X_ok.

Done. I have added that new pass as an additional commit, which partly reverts some of the earlier changes. So it might make sense to review the overall diff rather than individual commits on this occasion.

**Review by kroening — APPROVED** [review]

**Review by remi-delmas-3000 — APPROVED** [review]

File: regression/cbmc/realloc-should-not-free-on-failure-to-allocate/test.desc
Comment (remi-delmas-3000) [inline regression/cbmc/realloc-should-not-free-on-failure-to-allocate/test.desc:1]: is this meant to be in this commit ?

File: regression/goto-cc-goto-analyzer/instrument_preconditions_locations/test.desc
Comment (remi-delmas-3000) [inline regression/goto-cc-goto-analyzer/instrument_preconditions_locations/test.desc:1]: is this meant to be in this commit ?

**Review by remi-delmas-3000 — CHANGES_REQUESTED** [review]

**Review by remi-delmas-3000 — APPROVED** [review]

File: regression/cbmc/realloc-should-not-free-on-failure-to-allocate/test.desc
Comment (remi-delmas-3000) [inline regression/cbmc/realloc-should-not-free-on-failure-to-allocate/test.desc:1]: is this meant to be here ?

File: regression/goto-cc-goto-analyzer/instrument_preconditions_locations/test.desc
Comment (remi-delmas-3000) [inline regression/goto-cc-goto-analyzer/instrument_preconditions_locations/test.desc:1]: is this meant to be here ?

File: src/cbmc/cbmc_parse_options.cpp
Comment (remi-delmas-3000) [inline src/cbmc/cbmc_parse_options.cpp:409]: why is this new encoding switchable ? and what happens when it is switched off ?

File: src/goto-programs/rewrite_rw_ok.h
Comment (remi-delmas-3000) [inline src/goto-programs/rewrite_rw_ok.h:1]: Could you add some documentation of when this rewriting must take place or when it may be omitted ? I still have a hard time figuring out the lifecycle of these expressions. The ansi-c front end convers rw_ok expressions to the prophecy versions right away, and then these get eliminated right before symex with CBMC, but may or may not remain in the model when other tools are invoked ?

**Review by remi-delmas-3000 — CHANGES_REQUESTED** [review]

File: regression/cbmc/realloc-should-not-free-on-failure-to-allocate/test.desc
Comment (tautschnig) [reply regression/cbmc/realloc-should-not-free-on-failure-to-allocate/test.desc:1]: Yes: the intermediate version that did the expansion during goto-program conversion resulted in goto programs that have non-assert additional instructions (because of side-effect cleanup). This, in turn, affects what asserts get labelled as preconditions. So, really, it just changed the labels on some properties. Moving the expansion into a separate rewrite step undid this change.

File: regression/goto-cc-goto-analyzer/instrument_preconditions_locations/test.desc
Comment (tautschnig) [reply regression/goto-cc-goto-analyzer/instrument_preconditions_locations/test.desc:1]: Same as above: just a change in property labelling that was no longer in place after moving to a separate rewrite step.

File: src/cbmc/cbmc_parse_options.cpp
Comment (tautschnig) [reply src/cbmc/cbmc_parse_options.cpp:409]: When this is switched off, the goto program will contain `{r,w,rw}_ok_exprt`, and the analysis needs to be able to deal with these. The CHC-based approach, for example, does have its own way of handling them.

File: src/goto-programs/rewrite_rw_ok.h
Comment (tautschnig) [reply src/goto-programs/rewrite_rw_ok.h:1]: I have added another sentence to this file's documentation.

File: src/ansi-c/expr2c_class.h
Comment (NlightNFotis) [inline src/ansi-c/expr2c_class.h]: This is present in line 29 as well. Safe to assume this was a squashing/rebasing mistake? Or is there a particular reason for why this was added here?

File: src/util/pointer_expr.cpp
Comment (NlightNFotis) [inline src/util/pointer_expr.cpp]: As far as I can tell, both `conjunction` and the `and_exprt` constructor here should behave in a functionally equivalent manner.

If my understanding is correct, is there any particular reason for why the two similar methods use a different method for the same functionality? 

If not, does it make sense to refactor them to align their internals in a way that shows the functions' equivalence (from a mechanical standpoint, I can see they're adding different conjuncts to the end expression).

File: src/goto-programs/goto_clean_expr.cpp
Comment (NlightNFotis) [inline src/goto-programs/goto_clean_expr.cpp]: The comment here would be better located around line 449, if I am not mistaken?

Reading it here made me assume that this function would do the cleanup, whereas in reality it just marks expressions for cleanup. The actual rewriting happens later, around line 449 as far as I can tell, correct?

File: src/goto-instrument/goto_instrument_parse_options.cpp
Comment (NlightNFotis) [inline src/goto-instrument/goto_instrument_parse_options.cpp:1773]: I see only the `full-slice` option is rewriting the `rw` expressions.

Is there some specific reason the other slices don't want to display the same functionality? Apologies, but I couldn't tell the intent here just by reading the code.

**Review by NlightNFotis — COMMENTED** [review]
Some questions before I hit approve.

File: src/goto-programs/goto_clean_expr.cpp
Comment (NlightNFotis) [reply src/goto-programs/goto_clean_expr.cpp]: **EDIT**: Ignore my comment here - I saw it got removed in a later commit when this check was removed.

File: src/ansi-c/expr2c_class.h
Comment (NlightNFotis) [reply src/ansi-c/expr2c_class.h]: There's also the question of the commit message mentioning:

> expr2c/{r,w,rw}_ok: add a dedicated method

But I don't see any method being added here. Is the commit message in sync with the intended change here, or do I misunderstand the intent?

Comment (NlightNFotis) [reply src/ansi-c/expr2c_class.h]: **EDIT**: Yeah, when I viewed all of the files in the unified file changes tab, it seems that the change isn't there, meaning that the addition was removed at another commit and I missed it.

That probably means that the whole of the first commit is redundant, if I'm not mistaken.

Comment (tautschnig) [reply src/ansi-c/expr2c_class.h]: Thank you for this careful review! Indeed, the bulk of the first commit was already merged in #7460, so I just need to drop this commit.

File: src/util/pointer_expr.cpp
Comment (tautschnig) [reply src/util/pointer_expr.cpp]: Oh, I have no idea why I wasn't consistent here -- these are indeed equivalent, except `conjunction` will have an extra layer of indirection. Now consistently using `and_exprt` here.

File: src/goto-instrument/goto_instrument_parse_options.cpp
Comment (tautschnig) [reply src/goto-instrument/goto_instrument_parse_options.cpp:1773]: Other forms of slicing don't use expression-level granularity. I have updated the commit message to include this information.

**Review by remi-delmas-3000 — APPROVED** [review]

**Comment by feliperodri** [comment]
@NlightNFotis could you give a final look?

**Comment by TGWDB** [comment]
> @NlightNFotis could you give a final look?

If you specifically want Fotis it will take a couple of weeks, but others here can take a look much sooner than that.

**Comment by tautschnig** [comment]
> > @NlightNFotis could you give a final look?
> 
> If you specifically want Fotis it will take a couple of weeks, but others here can take a look much sooner than that.

I believe I have addressed all of Fotis's comments, so it should likely be just any code owner for the incremental SMT2 back-end. Will update assignments accordingly.

**Review by thomasspriggs — APPROVED** [review]
Approving for the incremental SMT2 changes.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp
Comment (tautschnig) [reply src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: Thank you, has been fixed.

File: src/solvers/smt2_incremental/convert_expr_to_smt.cpp
Comment (tautschnig) [reply src/solvers/smt2_incremental/convert_expr_to_smt.cpp]: Fixed according to your suggestion.

**Review by martin-cs — APPROVED** [review]
I believe that the changes have been implemented correctly and do not raise any issues in the parts of the code that I am "owner" for.

I am as yet unsure why this is needed or why this is a good design to meet that need.  However, judging by the comment history, this has been considered and it is something that I should resolve with a chat with @tautschnig rather than via github.

**Comment by tautschnig** [comment]
> I am as yet unsure why this is needed or why this is a good design to meet that need. However, judging by the comment history, this has been considered and it is something that I should resolve with a chat with @tautschnig rather than via github.

Let me try to clarify publicly nonetheless (the why-is-this-needed part):
1) Not all middle-ends agree on how `__CPROVER_{r,w,rw}_ok` ought to be processed. Notably, the `cprover` tool has its own way of handling these. So we shouldn't prematurely expand them in the C front-end.
2) Early expansion leads to redundant checks, which result in a performance penalty. See #7378 (and related issues linked from there).
3) In the status before the changes in this PR, this early expansion was necessary (as not all middle-ends were able to cope with these expressions), and this was silently done as part of the property instrumentation code (even when no property checks were requested). This is surprising behaviour.

**Review by peterschrammel — APPROVED** [review]

**Pull request merged**
